rust-src/concordium_base/src/id/sigma_protocols/com_ineq.rs

```diff
@@ -0,0 +1,163 @@
+//! inequality for committed value and public value". This protocol enables one
+//! to prove that a committed value is not equal to a public, without revealing
+//! the value.
+
+use super::{
+    com_mult::{ComMult, ComMultSecret, Witness as ComMultWitness},
+    common::{prove as sigma_prove, verify as sigma_verify, SigmaProof},
+};
+use crate::{
+    common::*,
+    curve_arithmetic::Curve,
+    pedersen_commitment::{Commitment, CommitmentKey, Randomness, Value},
+    random_oracle::RandomOracle,
+};
+use ff::Field;
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
+pub struct Witness<C: Curve> {
+    // The witness consists of a com_mult witness and a commitment.
+    com_mult_witness: SigmaProof<ComMultWitness<C>>,
+    aux_com:          Commitment<C>,
+}
+
+/// Function for proving that a committed value `value` is different from
+///  `pub_value`. The parameters are
+/// - `com_key` - the commitment key used to commit to the value
+/// - `value` - the value claimed to be inside the commitment
+/// - `value_tilde` - the randomness used to commit
+/// - `pub_value` - the public value claimed to be different from `value`
+
+pub fn prove_com_ineq<R: rand::Rng, C: Curve>(
+    com_key: &CommitmentKey<C>,
+    value: &Value<C>,
+    value_tilde: Randomness<C>,
+    pub_value: C::Scalar,
+    csprng: &mut R,
+) -> Option<Witness<C>> {
+    let mut transcript = RandomOracle::domain(b"NonEqualityProof");
+
+    let c = com_key.hide(&value, &value_tilde);
+    transcript.append_message(b"g", &com_key.g);
+    transcript.append_message(b"h", &com_key.h);
+    transcript.append_message(b"public commitment", &c);
+    transcript.append_message(b"public value", &pub_value);
+
+    // compute value - pub_value
+    let mut diff = pub_value;
+    diff.negate();
+    diff.add_assign(value);
+
+    // compute the inverse of (value - pub_value)
+    let diff_inv = diff.inverse()?;
+
+    // generate commitment cmm_1 to `diff` using the given randomness `value_tilde`;
+    // this allows the verifier to compute cmm_1 from public commitment
+    // c=g^{value} h^{value_tilde} for com_key=(g, h).
+    let diff_val = Value::new(diff);
+    let diff_inv_val = Value::new(diff_inv);
+    let cmm_1 = com_key.hide(&diff_val, &value_tilde);
+    let (cmm_2, r_2) = com_key.commit(&diff_inv_val, csprng);
+    //  Compute a commitment to 1. Alternatively, one could use cmm_3 =
+    // com_key.hide(one, zero).
+    let cmm_3 = Commitment(com_key.g);
+
+    // This is the com-mult proof in section 9.2.10 from the Bluepaper.
+    let prover = ComMult {
+        cmms:    [cmm_1, cmm_2, cmm_3],
+        cmm_key: *com_key,
+    };
+
+    let secret = ComMultSecret {
+        values: [diff_val, diff_inv_val],
+        rands:  [value_tilde, r_2, Randomness::<C>::zero()],
+    };
+
+    let partial_proof = sigma_prove(&mut transcript, &prover, secret, csprng);
+    let partial_proof = partial_proof.unwrap();
```

Comment:
I think `unwrap` is not good practice. We should either handle the error here or propagate it.